### DISCUSSION TITLE
Making and reviewing CSS changes in the web client

### DISCUSSION THREAD

**Discussion by GaryElshaw** [transmission/transmission#5024] [General]
What's the best way of going about making and reviewing changes to the CSS (and for that matter JS) of the web client?

**Comment by ckerr** [comment]
Thanks for asking!

- If you'd like suggestions, please consider the bugs in [this list](https://github.com/transmission/transmission/issues?q=is%3Aissue+is%3Aopen+label%3A%22pr+welcome%22+label%3A%22scope%3Aweb%22+). The way to an overworked maintainer's heart is to make the issue list smaller :smile_cat: 

- If it's an easy change for you to make, then a PR with A/B screenshots to demo the change ought to work.

- But If it's something that takes a little time to write, maybe open an issue first to see if other people on the project like the idea.

- The actual mechanics of making changes are

    Terminal 1: start webpack running 
    ```sh
    $ cd transmission/web/
    $ npm install
    $ npm run dev
   ```

    Terminal 2: start Transmission so that it can serve you transmission-web
    ```sh
    $ cd  transmission/build/
    $ TRANSMISSION_WEB_HOME="$PWD/../web/public_html" path/to/local/build/of/transmission
   ```
    (note: that previous step assumes "build" and "web" are sibling directories. If they're not, change the value of `TRANSMISSION_WEB_HOME` appropriately)
  
   Browser: load transmission-web being served by the Transmission in terminal 2

    Terminal 3 or your IDE or choice: edit the files in `transmission/web/src` and/or `transmission/web/assets`. When you save a file back to disk, webpack will regenerate the webpack file (you can confirm this by watching terminal 1) and when it's done its work, then you can force-reload in the browser to see your changes. **Remember to [hard refresh](https://fabricdigital.co.nz/blog/how-to-hard-refresh-your-browser-and-clear-cache)** or you will be frustrated wondering why your changes aren't doing anything  :smile: 

- When you're happy with your changes and are going to commit them to your branch, remember to _not_ commit changes to `web/public_html/` files. Those are auto-generated and get updated by a GitHub action.

Thanks again for asking!

**Comment by dareiff** [comment]
I have the beginnings of a README for the web directory around here somewhere. I'll work on that and let someone decide if it's worth adding.
